src/decode/util.rs

```diff
@@ -35,33 +34,32 @@ pub fn flush_zero_padding<R: io::BufRead>(input: &mut R) -> io::Result<bool> {
 }
 
 // A Read computing a digest on the bytes read.
-pub struct HasherRead<'a, R, H>
+pub struct DigestRead<'a, 'b, R, S>
 where
     R: 'a + io::Read,
-    H: 'a + hash::Hasher,
+    S: crc::Width,
 {
-    read: &'a mut R,   // underlying reader
-    hasher: &'a mut H, // hasher
+    read: &'a mut R,                    // underlying reader
+    digest: &'a mut crc::Digest<'b, S>, // hasher
 }
 
-impl<'a, R, H> HasherRead<'a, R, H>
+impl<'a, 'b, R, S> DigestRead<'a, 'b, R, S>
 where
     R: io::Read,
-    H: hash::Hasher,
+    S: crc::Width,
 {
-    pub fn new(read: &'a mut R, hasher: &'a mut H) -> Self {
-        Self { read, hasher }
+    pub fn new(read: &'a mut R, digest: &'a mut crc::Digest<'b, S>) -> Self {
+        Self { read, digest }
     }
 }
 
-impl<'a, R, H> io::Read for HasherRead<'a, R, H>
+impl<'a, 'b, R> io::Read for DigestRead<'a, 'b, R, u32>
```

Comment:
Because of how [`impl Digest`](https://docs.rs/crc/latest/crc/struct.Digest.html) is defined there's no way to define it once for all possible types.

In practice we only need `u32` for `impl Read` for usage in lzma-rs, so it's only implemented for `u32` here.

Reply:
In the end, due to this limitation, the old `HasherRead` is not as generic anymore, and only works for `digest == CRC32`. I see two ways to resolve this design conflict:

1. Keep most of the original implementation, but replace `hash::Hasher` by a new `Digest` trait that only contains `fn update(&mut self, bytes: &[u8])` (without the `finalize` part) - and add an `impl Digest for crc::Digest<_>`. That way it's still compatible with all the CRC sizes, as well other hash algorithms.
2. Remove the genericity over the digest type, and rebrand the `HasherRead` as `Crc32DigestRead` (with only genericity over the `io::Read`).

I'd lean towards solution (1) to minimize code diff and to unlock potential support for other digest algorithms, but (2) would be fine as well.

Reply:
`DigestRead`/`DigestWrite` is still generic over other CRC sizes, there just needs to be an impl for sizes other than `u32` which I've omitted for YAGNI reasons; adding sizes other than CRC32 is trivial where the implementation is exactly the same. If you prefer, we can impl it for all known CRC sizes with a macro.  What it is no longer generic over is however, all implementations of `Hasher` that are not CRC. Again, for YAGNI reasons I don't think there is currently a strong case for (1).

Instead, I've went with (2) and clarified the naming as `CrcDigestRead`/`CrcDigestWrite`.